---
title: "250312_GTExAnalysis"
author: "Irika Sinha"
date: "2025-03-14"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Purpose

The purpose of this script is to take the file with manually identified cryptic exon coordinates and identify inclusion in GTEx tissues of CEs

## Packages

```{r packages}
CRANpackages <- c("tidyverse", "ggplot2", "ggpubr", "ggstats","BiocManager","ggrepel", "rstatix","pheatmap","ggridges")

for(i in CRANpackages){
  if(!(i %in% rownames(installed.packages()))){
    install.packages(pkgs = i)
  }
  library(package = i, character.only = T, quietly = T)
  rm(i)
}

rm(CRANpackages)

```

## Functions

```{r functions}
regExSamples <- function(df){
  #create df for samples + count
  allSamples <- df$samples
  allSamples<- unlist(strsplit(allSamples,",")) #separate character into string of characters
  allSamples <- allSamples[allSamples!=""] #remove blank samples
  
  #RegEx information: https://www.journaldev.com/36776/regular-expressions-in-r
  sampleID <- gsub(":[^:]+$","",allSamples) #replace everything after and including colon with nothing
  sampleCount <- gsub("^[^:]+:", "",allSamples) #replace everything before and including colon with nothing
  sampleCount <- as.numeric(sampleCount)
  sampleDF <- data.frame(sampleID, sampleCount)
  colnames(sampleDF)=c("sampleID","junctionCount")
  
  return(sampleDF)
}
calcPSI <- function(chr,incA,incB,incC = NA,incD = NA,excA,excD,genome, totalCountMin = 15){
  #Download snaptron junction info around region bounded by exclusion junction
  url <- paste0("https://snaptron.cs.jhu.edu/",genome,"/snaptron?regions=chr",
                chr,":",excA-100,"-",excD + 100)
  options(timeout=300)
  snaptronQuery <- read.csv(url, sep="\t",quote = "")
  
  #Create Count DFs
  if(!is.na(incC) & !is.na(incD)){
    junctions <- list(inc1 = c(incA,incB), 
                      inc2 = c(incC, incD),
                      exc = c(excA, excD))
    
    for(i in 1:3){
      junctionRow <- which(snaptronQuery$start==junctions[[i]][1] & snaptronQuery$end==junctions[[i]][2]) #determine relevant row in Snaptron file
      junctionDF <- snaptronQuery[junctionRow,]
      junctionDF <- regExSamples(junctionDF)
      colnames(junctionDF)[2] = paste0(names(junctions)[i],"_count")
      assign(paste0(names(junctions)[i],"_count"),junctionDF)
    }
    
    countdf <- merge(inc1_count,inc2_count,sort = T, all = T) #include all samples
    countdf <- merge(countdf,exc_count,sort = T, all = T) #include all samples
    countdf[is.na(countdf)] = 0 #any junctions with no counts of a type will be set to 0
    
    #Add PSI values
    countdf <- countdf %>% mutate(totalExonInc1 = inc1_count + exc_count) %>% mutate(PSI_Inc1 = inc1_count/totalExonInc1)
    countdf <- countdf %>% mutate(totalExonInc2 = inc2_count + exc_count) %>% mutate(PSI_Inc2 = inc2_count/totalExonInc2)
    countdf$PSI_Inc1 <- round(countdf$PSI_Inc1*100,2)
    countdf$PSI_Inc2 <- round(countdf$PSI_Inc2*100,2)
    
    #Set PSI to NA if not enough counts of junction to indicate real measurements
    countdf$PSI_Inc1[which(countdf$totalExonInc1 < totalCountMin)] = 0
    countdf$PSI_Inc2[which(countdf$totalExonInc2 < totalCountMin)] = 0
    
    #Generate average
    countdf <- countdf %>% mutate(avgPSI = rowMeans(select(countdf, c(PSI_Inc1,PSI_Inc2)), na.rm = TRUE))
    
    #rename sampleID columns to numeric
    countdf$sampleID <- as.numeric(countdf$sampleID)
    
  } else{
    junctions <- list(inc1 = c(incA,incB), 
                      exc = c(excA, excD))
    
    for(i in 1:2){
      junctionRow <- which(snaptronQuery$start==junctions[[i]][1] & snaptronQuery$end==junctions[[i]][2]) #determine relevant row in Snaptron file
      junctionDF <- snaptronQuery[junctionRow,]
      junctionDF <- regExSamples(junctionDF)
      colnames(junctionDF)[2] = paste0(names(junctions)[i],"_count")
      assign(paste0(names(junctions)[i],"_count"),junctionDF)
    }
    
    countdf <- merge(inc1_count,exc_count,sort = T, all = T) #include all samples
    countdf[is.na(countdf)] = 0 #any junctions with no counts of a type will be set to 0
    
    #Add PSI values
    countdf <- countdf %>% mutate(totalExonInc1 = inc1_count + exc_count) %>% mutate(PSI_Inc1 = inc1_count/totalExonInc1)
    countdf$PSI_Inc1 <- round(countdf$PSI_Inc1*100,2)
    
    #Set PSI to NA if not enough counts of junction to indicate real measurements
    countdf$PSI_Inc1[which(countdf$totalExonInc1 < totalCountMin)] = NA
    countdf$avgPSI = countdf$PSI_Inc1
    
    #rename sampleID columns to numeric
    countdf$sampleID <- as.numeric(countdf$sampleID)
  }
  
  return(countdf)
}
samplesInfo <- function(genome, sampleIDs){
  #Download samples to local folder: "curl https://snaptron.cs.jhu.edu/data/srav1m/samples.tsv >srav1m_samples.tsv"
  fname <- paste0(genome, "_samples.tsv")
  sampleInfo <- read_tsv(fname, lazy = TRUE,quote = "")
  sampleInfo <- sampleInfo %>% select(c("rail_id","external_id","study","study_title",
                                        "library_layout","sample_description","study_abstract",
                                        "study_description","sample_name","sample_title")) %>% filter(rail_id %in% sampleIDs) 
  return(sampleInfo)
}
```

## Files
```{r fileUpload}
setwd("~/Library/CloudStorage/OneDrive-JohnsHopkins/WongLing/Scripts/250314_SnaptronRevisions/SnapMine_Paper/Scripts/Revisions/")
# Create info file
info <- read_delim("250312_info.csv")[-12]

#Run across info file to populate PSI df
df <- data.frame("sampleID"=matrix(ncol = 1, nrow = 0))

for(i in 1:nrow(info)){
  dftemp <- calcPSI(chr = info$chr[i], incA = info$hg38_inc1_A[i], incB = info$hg38_inc1_B[i],
                    incC = info$hg38_inc2_C[i], incD = info$hg38_inc2_D[i], 
                    excA = info$hg38_exc_A[i], excD = info$hg38_exc_D[i], genome = "gtexv2")
  dftemp <- dftemp %>% select(sampleID,starts_with("PSI"),ends_with("PSI")) %>% mutate_all(~ifelse(is.nan(.), NA, .))
  colnames(dftemp)[2:ncol(dftemp)] = paste0(info$gene[i],"__",colnames(dftemp)[2:ncol(dftemp)])
  
  df <- merge(df, dftemp,all = T)
  print(i)
  rm(dftemp,i)
}

#write.csv(df,"250314_PSI_GTEx.csv")
df <- read.csv("250314_PSI_GTEx.csv")[-1]

#Incorporate metadata and save CSV file for junctions
meta <- read_delim("gtexv2_samples.tsv") %>% dplyr::select(c("rail_id","study", "external_id", "SUBJID","SAMPID", "SEX","AGE","DTHHRDY","SMTS","SMTSD"))
df <- merge(meta,df %>% dplyr::select(c(sampleID,contains("avgPSI"))), 
            by.x = "rail_id", by.y = "sampleID")

# Combine CE with multiple isoform
df <- df %>% mutate(AGRN__avgPSI = max(AGRN_long__avgPSI,AGRN_short__avgPSI), 
                    ATG4__avgPSIB = max(ATG4B_long__avgPSI,ATG4B_short__avgPSI),
                    KALRN__avgPSI = max(KALRN_long__avgPSI,KALRN_short__avgPSI),
                    MYO18A__avgPSI = max(MYO18A__avgPSI,MYO18A_alt__avgPSI),
                    UNC13A__avgPSI = max(UNC13A__avgPSI,UNC13A_alt__avgPSI)) %>%
  relocate(AGRN__avgPSI, .before = AGRN_long__avgPSI) %>%
  relocate(ATG4__avgPSIB, .before = ATG4B_long__avgPSI) %>%
  relocate(KALRN__avgPSI, .before = KALRN_long__avgPSI) %>%
  relocate(MYO18A__avgPSI, .before = MYO18A_alt__avgPSI) %>%
  relocate(UNC13A__avgPSI, .before = UNC13A_alt__avgPSI) %>%
  dplyr::select(!c(AGRN_long__avgPSI,AGRN_short__avgPSI,ATG4B_long__avgPSI,ATG4B_short__avgPSI,
                   KALRN_long__avgPSI, KALRN_short__avgPSI, MYO18A_alt__avgPSI,
                   UNC13A_alt__avgPSI))

#write.csv(df, "250314_Cryptic_GTExAllCEPSI_meta.csv")

```

### Find GTEx sample IDs for individuals with undiagnosed TDP-43 pathology

Individuals with undiagnosed TDP-43 will have multiple cryptic exons from the panel of cryptic exons. As a conservative cut-off, individuals with >=5 cryptic exons in a single tissue will be excluded from further analyses as potentially having TDP-43 pathology.

```{r TDPundiagnosed}
setwd("~/Library/CloudStorage/OneDrive-JohnsHopkins/WongLing/Scripts/250314_SnaptronRevisions/SnapMine_Paper/Scripts/Revisions/")
df <- read.csv("250314_Cryptic_GTExAllCEPSI_meta.csv", row.names = 1)
subjects <- meta %>% dplyr::select(SUBJID,SEX, AGE, DTHHRDY) %>% distinct()

# Subset GTEx PSI df to samples with >5% avg junction PSI
df_posCE <- df %>% pivot_longer(11:ncol(df), names_to = "CE", values_to = "avgPSI") %>%
  mutate(CE = gsub(x = CE, pattern = "__avgPSI", replacement = ""),
         avgPSI = ifelse(is.na(avgPSI),0,avgPSI),
         cutoff = ifelse(avgPSI > 5, "Above", "Below")) 

# Find samples with CE in >1 tissue
nCE_Tissue <- df_posCE %>% filter(cutoff == "Above", !is.na(SUBJID)) %>% 
  dplyr::select(SUBJID,SEX, AGE, DTHHRDY, SMTS, CE) 

# Visualize # of CE in each tissue for each individual with at least one - add 0s
nCE_Tissue_Indiv <- nCE_Tissue %>%  group_by(SUBJID,SMTS) %>%
  summarize(UniqueCE = n()) %>%
  left_join(y = subjects, by = join_by("SUBJID" == "SUBJID")) %>%
  mutate(SEX = ifelse(SEX == 1, "male", ifelse(SEX == 2, "female", "unknown"))) %>%
  pivot_wider(names_from = SMTS, values_from = UniqueCE, values_fill = 0) %>%
  pivot_longer(5:34, names_to = "SMTS", values_to = "UniqueCE")
  
nCE_Tissue_Indiv %>% ggplot(aes(x = SMTS, y = UniqueCE)) + 
  geom_hline(yintercept = 3, linetype = "dotted")+
  geom_point(position = position_jitter(width = 0.2, height = 0.5), size = 2, color = "black")+
  geom_boxplot(outliers = F, alpha = 0.9, fill = "#AEAEAE") + 
  theme_classic() +theme(text = element_text(size = 14, color = "black"), axis.ticks = 
                           element_blank(),
                         axis.text.x = element_text(angle = 90, hjust = 1)) +
  ylab("# Unique Cryptic Exon per Individual") + xlab("") + 
  scale_y_continuous(expand = c(0,0),  limits = c(0,7))  

ggsave("250314_finalGTEX_AllPosCE.png",device = "png",
       width = 50, height = 30, units = "cm",limitsize = FALSE)

# Individuals with TDP path: >5 CE in a single tissue
undiagnosedPath_Indiv <- nCE_Tissue_Indiv %>% filter(UniqueCE >= 5) %>% 
  arrange(-UniqueCE)
#write.csv(undiagnosedPath_Indiv, "250314_undiagnosedPath_Indiv_GTEx.csv")

nCE_Tissue_Indiv %>% filter(SUBJID %in% undiagnosedPath_Indiv$SUBJID) %>%
  pivot_wider(names_from = SMTS, values_from = UniqueCE, values_fill = 0) %>%
  pivot_longer(5:34, names_to = "SMTS", values_to = "UniqueCE") %>%
  ggplot(aes(x = SMTS, y = UniqueCE)) + 
  geom_hline(yintercept = 3, linetype = "dotted")+
  geom_point(position = position_jitter(width = 0.2, height = 0.5), size = 2, color = "black")+
  geom_boxplot(outliers = F, alpha = 0.9, fill = "#AEAEAE") + 
  theme_classic() +theme(text = element_text(size = 14, color = "black"), axis.ticks = 
                           element_blank(),
                         axis.text.x = element_text(angle = 90, hjust = 1)) +
  ylab("# Unique Cryptic Exon per Individual") + xlab("") + 
  scale_y_continuous(expand = c(0,0),  limits = c(0,19)) 

ggsave("250314_finalGTEX_AllPosCE_TDPpos.png",device = "png",
       width = 50, height = 30, units = "cm",limitsize = FALSE)

# Max PSI in brain
df_TDPpos_CEmax_brain <- df_posCE %>%  
  filter(SMTS == "Brain") %>%
  dplyr::select(SUBJID, SMTS, CE, avgPSI) %>%
  group_by(SUBJID, SMTS, CE) %>% summarize(maxPSI = max(avgPSI)) %>%
  filter(SUBJID %in% undiagnosedPath_Indiv$SUBJID) %>%
  mutate(maxPSI = ifelse(maxPSI>20, 20, maxPSI)) 

df_TDPpos_CEmax_brain %>% 
  ggplot(aes(y = CE, x = SUBJID, fill = maxPSI)) + geom_tile(color = "#362417") + 
  scale_fill_gradient2(low="white",high="#362417",mid="#C12327",midpoint=10)+
  scale_y_discrete(expand=c(0, 0),)+ scale_x_discrete(expand=c(0, 0)) +
  theme_classic() + ylab("") + xlab("") + scale_y_discrete(limits=rev) +
  theme(text = element_text(size = 14, color = "black"),axis.line=element_blank(), axis.ticks  =
          element_blank(),
        axis.text.x = element_text(angle=90, vjust=1,hjust=1,family="Helvetica"))
ggsave("250314_finalGTEX_brainMaxPSI_TDPpath.png",device = "png",
       width = 30, height = 30, units = "cm",limitsize = FALSE)
# write.csv(df_TDPpos_CEmax_brain, "250314_TDPpos_CrypticBrain_GTExPSI_meta.csv")
```

## GTEx cryptic exon inclusion in non-TDP-43 pathology individuals

```{r Fig2}
setwd("~/Library/CloudStorage/OneDrive-JohnsHopkins/WongLing/Scripts/250314_SnaptronRevisions/SnapMine_Paper/Scripts/Revisions/")
wd <- "~/Library/CloudStorage/OneDrive-JohnsHopkins/WongLing/Scripts/250314_SnaptronRevisions/SnapMine_Paper/Scripts/Revisions/"

df <- read.csv("250314_Cryptic_GTExAllCEPSI_meta.csv", row.names = 1)
tdpPos_df <- read.csv("250314_undiagnosedPath_Indiv_GTEx.csv", row.names = 1)
tdpPos_df <- unique(tdpPos_df$SUBJID)
df <- df %>% filter(!(SUBJID %in% tdpPos_df)) 
#write.csv(df, "250314_Cryptic_GTExPSI_noTDP_meta.csv")
df <- read.csv("250314_Cryptic_GTExPSI_noTDP_meta.csv")[-1]

# Create a df for plotting
df_vis <- df %>% pivot_longer(11:32, names_to = "CE", values_to = "avgPSI") %>%
  mutate(CE = gsub(x = CE, pattern = "__avgPSI", replacement = ""),
         avgPSI = ifelse(is.na(avgPSI),0,avgPSI)) 

genes = c("AGRN", "ATG4B", "EPB41L4A","G3BP1", "HDGFL2", "MYO18A", "PFKP", "RANBP1",
          "STMN2", "UNC13A", "UNC13B")

# Main figure - 2A
df_vis %>% filter(CE %in% genes) %>%
  ggplot(aes(x = CE, y = avgPSI)) +  
  geom_jitter(alpha = 0.8, color = "#A365A9",na.rm = T,width = 0.2, height = 0) +
  geom_boxplot(alpha = 0.5,width=0.5, fill = "white",
               position=position_dodge(width=0.2),
               outlier.shape = NA,lwd=0.3)+ 
  geom_hline(yintercept=0,linetype=1, color = "black") +
  theme_classic() + 
  theme(axis.line=element_line(), 
        axis.title.x = element_blank(),
        axis.text.x = element_text(angle=90, vjust=1, hjust=1,family="Helvetica"),
        axis.ticks.x  = element_blank())+ 
  scale_y_continuous(expand = c(0,0),  limits = c(0,100))
ggsave(paste0(wd,"250314_finalGTEX_2A_CE_PSI.png"),device = "png",width = 20, height = 20, units = "cm",limitsize = FALSE)

# Full set
df_vis %>% 
  mutate(avgPSI = ifelse(avgPSI >=20, 20, avgPSI)) %>%
  ggplot(aes(x = SMTS, y = avgPSI)) +  
  geom_jitter(alpha = 0.8, aes(color = avgPSI),na.rm = T,width = 0.2, height = 0) +
  scale_color_gradient2(low="#86C7ED",mid="#A365A9",high="#C12327",midpoint = 10)+
  geom_boxplot(alpha = 0.5,width=0.5, fill = "white",
               position=position_dodge(width=0.2),
               outlier.shape = NA,lwd=0.3)+ 
  geom_hline(yintercept=0,linetype=1, color = "black") +
  facet_wrap(vars(CE),ncol=1,strip.position = "right")+
  theme_classic() + 
  theme(strip.background = element_rect(fill = "#4B4E6D"), 
        strip.placement = "outside",
        strip.text = element_text(color="white",family="Helvetica"),
        strip.text.y = element_text(angle=0),
        axis.line=element_line(), 
        axis.title.x = element_blank(),
        axis.text.x = element_text(angle=90, vjust=1, hjust=1,family="Helvetica"),
        axis.ticks.x  = element_blank())+ 
  scale_y_continuous(expand = c(0,0),  limits = c(0,21))
ggsave(paste0(wd,"250314_finalGTEX_AllCE_PSI.png"),device = "png",width = 20, height = 20, units = "cm",limitsize = FALSE)

# Main figure - 2B
df_vis %>% filter(CE %in% genes) %>%
  mutate(avgPSI = ifelse(avgPSI >=20, 20, avgPSI)) %>%
  ggplot(aes(x = SMTS, y = avgPSI)) +  
  geom_jitter(alpha = 0.8, aes(color = avgPSI),na.rm = T,width = 0.2, height = 0) +
  scale_color_gradient2(low="#86C7ED",mid="#A365A9",high="#C12327",midpoint = 10)+
  geom_boxplot(alpha = 0.5,width=0.5, fill = "white",
               position=position_dodge(width=0.2),
               outlier.shape = NA,lwd=0.3)+ 
  geom_hline(yintercept=0,linetype=1, color = "black") +
  facet_wrap(vars(CE),ncol=1,strip.position = "right")+
  theme_classic() + 
  theme(strip.background = element_rect(fill = "black"), 
        strip.placement = "outside",
        strip.text = element_text(color="black",family="Helvetica"),
        axis.line=element_line(), 
        axis.title.x = element_blank(),
        axis.ticks.x  = element_blank(),
        axis.text.x = element_text(angle=90, vjust=1, hjust=1,family="Helvetica"))+ 
  scale_y_continuous(expand = c(0,0),  limits = c(0,21))
ggsave(paste0(wd,"250314_finalGTEX_2B_CE_PSI.png"),device = "png",width = 20, height = 20, units = "cm",limitsize = FALSE)

## Main figure - 2C
df_posPct <- df_vis %>% #filter(CE %in% genes) %>%
  mutate(cutoff = ifelse(avgPSI > 5, "Above", "Below"), 
         ID = paste0(SMTS,"_",CE)) 
df_posPct <- data.frame(table(df_posPct$ID, df_posPct$cutoff)) %>% 
  pivot_wider(names_from = "Var2", values_from = "Freq", values_fill = 0) %>%
  mutate(pct = round(100 * Above/(Above+Below),1))
df_posPct <- df_posPct %>% separate("Var1", into = c("tissue", "CE"), sep = "_")
  
df_posPct %>% filter(CE %in% genes) %>%
  mutate(pct = ifelse(pct >=20, 20, pct)) %>%
  ggplot(aes(y = CE, x = tissue, fill = pct)) + geom_tile(color = "#362417") + 
  scale_fill_gradient2(low="white",high="#362417",mid="#C12327",midpoint=10)+
  scale_y_discrete(expand=c(0, 0),)+ scale_x_discrete(expand=c(0, 0)) +
  theme_classic() + ylab("") + xlab("") + scale_y_discrete(limits=rev) +
  theme(text = element_text(size = 14, color = "black"),
        axis.line=element_blank(),
        axis.ticks= element_blank(),
        axis.text.x = element_text(angle=90, vjust=1,hjust=1,family="Helvetica"))
ggsave(paste0(wd,"250314_finalGTEX_2C_CE_PSI.png"),device = "png",width = 20, height = 20, units = "cm",limitsize = FALSE)

```
